runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java

```diff
@@ -199,4 +208,67 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
       return ReplacementOutputs.tagged(outputs, newOutput);
     }
   }
+
+  @VisibleForTesting
+  static class StreamingShardedWriteFactory<UserT, DestinationT, OutputT>
+          implements PTransformOverrideFactory<
+          PCollection<UserT>, WriteFilesResult<DestinationT>,
+          WriteFiles<UserT, DestinationT, OutputT>> {
+    FlinkPipelineOptions options;
+
+    StreamingShardedWriteFactory(PipelineOptions options) {
+      this.options = options.as(FlinkPipelineOptions.class);
+    }
+
+    private static final Integer DEFAULT_MAX_PARALLELISM = 128;
+
+    @Override
+    public PTransformReplacement<PCollection<UserT>, WriteFilesResult<DestinationT>>
+    getReplacementTransform(
+            AppliedPTransform<
+                    PCollection<UserT>, WriteFilesResult<DestinationT>,
+                    WriteFiles<UserT, DestinationT, OutputT>>
+                    transform) {
+      // By default, if numShards is not set WriteFiles will produce one file per bundle. In
+      // streaming, there are large numbers of small bundles, resulting in many tiny files.
+      // Instead we pick parallelism * 2 to ensure full parallelism, but prevent too-many files.
+      Integer jobParallelism = options.getParallelism();
+
+      if (jobParallelism == -1) {
```

Comment:
job parallelism should be guaranteed by now to by > 0. There is a test for it also in `FlinkExecutionEnvironmentsTest`.